python/ray/experimental/raysort/main.py

```diff
@@ -190,146 +208,270 @@ def _dummy_sort_and_partition(part: np.ndarray,
 
 
-@ray.remote
-def reducer(reducer_id: PartId, *blocks: List[ray.ObjectRef]) -> PartitionInfo:
-    logging_utils.init()
-    task_id = f"R-{reducer_id} Reducer"
-    logging.info(f"{task_id} starting")
-    blocks = [np.copy(ray.get(block)) for block in blocks]
+@ray.remote(num_cpus=0, resources={"worker": 1e-3})
+def _write_data(path, data):
+    with open(path, "ab") as fout:
+        fout.write(data)
+
+
+def _merge_impl(args: Args,
+                M: int,
+                pinfo: PartInfo,
+                get_block: Callable[[int, int], np.ndarray],
+                skip_output=False,
+                remote_writer=True):
     merge_fn = _dummy_merge if args.skip_sorting else sortlib.merge_partitions
-    merger = merge_fn(blocks, args.reducer_batch_num_records)
-    if args.skip_output:
+    merger = merge_fn(M, args.reducer_output_chunk, get_block)
+
+    if skip_output:
         for datachunk in merger:
             del datachunk
-        logging.info(f"{task_id} done")
-        return None
     else:
-        pinfo = _make_partition_info(reducer_id, "output")
-        with open(pinfo.path, "wb") as fout:
+        if not remote_writer:
+            with open(pinfo.path, "wb") as fout:
+                for datachunk in merger:
+                    fout.write(datachunk)
+        else:
+            with open(pinfo.path, "wb"):
+                pass
+            task = None
             for datachunk in merger:
-                fout.write(datachunk)
-        logging.info(f"{task_id} done")
-        return pinfo
+                if task is not None:
+                    ray.get(task)
+                task = _write_data.remote(pinfo.path, datachunk)
+            if task is not None:
+                ray.get(task)
+    return pinfo
 
 
-@tracing_utils.timeit("sorting")
-def sort_main():
-    partitions = _load_manifest(constants.INPUT_MANIFEST_FILE)
+@ray.remote(num_cpus=0, resources={"worker": 1e-3})
+@tracing_utils.timeit("merge")
+def merge_mapper_blocks(args: Args, reducer_id: PartId, mapper_id: PartId,
+                        *blocks: List[np.ndarray]) -> PartInfo:
+    part_id = constants.merge_part_ids(reducer_id, mapper_id)
+    pinfo = _part_info(args, part_id, kind="temp")
+    M = len(blocks)
+
+    def get_block(i, d):
+        if i >= M or d > 0:
+            return None
+        return blocks[i]
+
+    _merge_impl(args, M, pinfo, get_block)
+    return pinfo
+
+
+@ray.remote(num_cpus=0, resources={"worker": 1e-3})
+@tracing_utils.timeit("reduce")
+def final_merge(args: Args, reducer_id: PartId,
+                *merged_parts: List[PartInfo]) -> PartInfo:
+    M = len(merged_parts)
+
+    @ray.remote(num_cpus=0, resources={"worker": 1e-3})
+    def _load_block_chunk(pinfo: PartInfo, d: int) -> np.ndarray:
+        return np.fromfile(
+            pinfo.path,
+            dtype=np.uint8,
+            count=args.reducer_input_chunk,
+            offset=d * args.reducer_input_chunk)
+
+    block_chunks = [
```

Comment:
Since this initial prototype is already bypassing the object store for the reduce step, I think it would be simpler here to just load the data into heap memory directly, even if it has to happen serially.

Later, we can measure the difference between this and using an object store-native approach for storing intermediate results in the reduce stage.

Reply:
Same as above.